pkg/registry/contractRegistry.go

```diff
@@ -0,0 +1,215 @@
+	if err := s.refreshData(); err != nil {
+		return err
+	}
+
+	go s.refreshLoop()
+
+	return nil
+}
+
+func (s *SmartContractRegistry) OnNewNodes() (<-chan []Node, CancelSubscription) {
+	return s.newNodesNotifier.register()
+}
+
+func (s *SmartContractRegistry) OnChangedNode(
+	nodeId uint16,
+) (<-chan Node, CancelSubscription) {
+	s.changedNodeNotifiersMutex.Lock()
+	defer s.changedNodeNotifiersMutex.Unlock()
+
+	notifier, ok := s.changedNodeNotifiers[nodeId]
+	if !ok {
+		notifier = newNotifier[Node]()
+		s.changedNodeNotifiers[nodeId] = notifier
+	}
+	return notifier.register()
+}
+
+func (s *SmartContractRegistry) GetNodes() ([]Node, error) {
+	s.nodesMutex.RLock()
+	defer s.nodesMutex.RUnlock()
+
+	nodes := make([]Node, len(s.nodes))
+	for idx, node := range s.nodes {
+		nodes[idx] = node
+	}
+	return nodes, nil
+}
+
+func (s *SmartContractRegistry) refreshLoop() {
+	ticker := time.NewTicker(s.refreshInterval)
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker.C:
+			if err := s.refreshData(); err != nil {
+				s.logger.Error("Failed to refresh data", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (s *SmartContractRegistry) refreshData() error {
+	fromContract, err := s.loadFromContract()
+	if err != nil {
+		return err
+	}
+
+	newNodes := []Node{}
+	for _, node := range fromContract {
+		existingValue, ok := s.nodes[node.NodeId]
+		if !ok {
+			// New node found
+			newNodes = append(newNodes, node)
+		} else if !node.Equals(existingValue) {
+			s.processChangedNode(node)
+		}
+	}
+
+	if len(newNodes) > 0 {
+		s.processNewNodes(newNodes)
+	}
+
+	return nil
+}
+
+func (s *SmartContractRegistry) processNewNodes(nodes []Node) {
+	s.logger.Info("processing new nodes", zap.Int("count", len(nodes)), zap.Any("nodes", nodes))
+	s.newNodesNotifier.trigger(nodes)
+
+	s.nodesMutex.Lock()
+	defer s.nodesMutex.Unlock()
+	for _, node := range nodes {
+		s.nodes[node.NodeId] = node
+	}
+}
+
+func (s *SmartContractRegistry) processChangedNode(node Node) {
+	s.nodesMutex.Lock()
+	defer s.nodesMutex.Unlock()
+	s.changedNodeNotifiersMutex.RLock()
+	defer s.changedNodeNotifiersMutex.RUnlock()
+
+	s.nodes[node.NodeId] = node
+	s.logger.Info("processing changed node", zap.Any("node", node))
+	if registry, ok := s.changedNodeNotifiers[node.NodeId]; ok {
+		registry.trigger(node)
+	}
+}
+
+func (s *SmartContractRegistry) loadFromContract() ([]Node, error) {
+	ctx, cancel := context.WithTimeout(s.ctx, CONTRACT_CALL_TIMEOUT)
+	defer cancel()
+	nodes, err := s.contract.AllNodes(&bind.CallOpts{Context: ctx})
+	if err != nil {
+		return nil, err
+	}
+	out := make([]Node, len(nodes))
+	for idx, node := range nodes {
+		out[idx] = convertNode(node)
+	}
+
+	return out, nil
+}
+
+func convertNode(rawNode abis.NodesNodeWithId) Node {
+	// Unmarshal the signing key.
+	// If invalid, mark the config as being invalid as well. Clients should treat the
+	// node as unhealthy in this case
+	signingKey, err := crypto.UnmarshalPubkey(rawNode.Node.SigningKeyPub)
```

Comment:
I think we should loudly surface an error instead of quietly excluding the node.

It's not expected for an invalid key to be written to the registry, and failing to read it here would imply some inconsistency in how the key is interpreted.

This would also be an argument for why we shouldn't let node operators change the http address themselves. It might be simpler for the DAO to do all validation before writing to the registry (or having the smart contract do the validation) so that everything in the registry can be treated as correct downstream.

Reply:
I can update it to at least log something loudly.

But I do think we need to treat the inputs in the smart contract as somewhat untrusted. Hopefully we'll pre-validate and this will never happen, but that's a human-driven manual process.